### PR TITLE
8287393: AArch64: Remove trampoline_call1

### PR DESCRIPTION
**Pull request by eastig**
`trampoline_call` can do dummy code generation to calculate the size of C2 generated code. This is done in the output phase. In [src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp#L1042](https://github.com/openjdk/jdk/blob/e0d361cea91d3dd1450aece73f660b4abb7ce5fa/src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp#L1042) Loom code needed to generate a trampoline call outside of C2 and without the output phase. This caused test crashes. The project Loom added `trampoline_call1` to workaround the crashes. 

This PR improves detection of C2 output phase which makes `trampoline_call1` redundant.

Tested the fastdebug/release builds:
- `'gtest`: Passed
- `tier1`...`tier2`: Passed

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287393](https://bugs.openjdk.org/browse/JDK-8287393): AArch64: Remove trampoline_call1

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9592/head:pull/9592` \
`$ git checkout pull/9592`

Update a local copy of the PR: \
`$ git checkout pull/9592` \
`$ git pull https://git.openjdk.org/jdk pull/9592/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9592`

View PR using the GUI difftool: \
`$ git pr show -t 9592`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9592.diff">https://git.openjdk.org/jdk/pull/9592.diff</a>

</details>

### CODE REVIEW

**Comment by eastig** [comment]
Andrew (@theRealAph), could you please have a look?

Thanks.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -624,13 +624,19 @@ address MacroAssembler::trampoline_call1(Address entry, CodeBuffer* cbuf, bool c
...
+    // trampoline stubs generated by Call LIR ops.
+    if (StubRoutines::aarch64::complete() &&
+        Thread::current()->is_Compiler_thread()) {
       // We don't want to emit a trampoline if C2 is generating dummy
       // code during its branch shortening phase.
       CompileTask* task = ciEnv::current()->task();
+      // We need to be in the output phase when we check the size of emitted
+      // code.
+      PhaseOutput* phase_output = Compile::current()->output();
       in_scratch_emit_size =
```
[the remark this answers is not in the document]
Comment (eastig) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:637]: Do you mean `Compile::current()`? 
`Compile::current()->output()` is checked in the expression for `in_scratch_emit_size` as `phase_output != NULL`.

[the remark this answers is not in the document]
Comment (eastig) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:637]: Hi @theRealAph,
I am sorry I did not get your comment. Could you please explain it?

Thanks,
Evgeny

[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:637]: The addition is
'PhaseOutput* phase_output = Compile::current()->output();'
then
'phase_output != NULL && phase_output->in_scratch_emit_size()'

so AFAICS `Compile::current()->output()` is now checked for null, where it was not before.

[the remark this answers is not in the document]
Comment (eastig) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:637]: Now I get it. Thank you.

I agree this looks suspicious. I could not recall why I added it.
Debugging helped me to find out.
During the parsing phase of C2 compilation `ciTypeFlow::StateVector::do_invoke` causes `LinkResolver::resolve_static_call` which now has the following code:
```
  if (resolved_method->is_continuation_enter_intrinsic()
      && resolved_method->from_interpreted_entry() == NULL) { // does a load_acquire
    methodHandle mh(THREAD, resolved_method);
    // Generate a compiled form of the enterSpecial intrinsic.
    AdapterHandlerLibrary::create_native_wrapper(mh);
  }
```
We generate a wrapper which is `nmethod` with trampoline calls.
As we are in the parsing phase the output is not created.
I can move `Compile::current()->output() != NULL` into the preceding IF and update the comment to the following:
```
Make sure this is code generation of a C2 compilation when Compile::current()->output() is not NULL.
C2 can generate native wrappers for the continuation enter intrinsic before code generation.
C1 allocates space only for trampoline stubs generated by Call LIR ops.
```

[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:637]: This is all rather complicated and obscure. It seems to me that passing a bool `check_emit_size` is exactly what we should do: it's more explicit and helps the reader.

[the remark this answers is not in the document]
Comment (eastig) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp]: I've restored `check_emit_size` and created an assert to guard it is properly used.
I'll remove `cbuf` by fixing: https://bugs.openjdk.org/browse/JDK-8287394

[the remark this answers is not in the document]
Comment (eastig) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp]: Hi Andrew,
Any comments on the latest version?

**Review by theRealAph — APPROVED** [review]

**Comment by dholmes-ora** [comment]
This is crashing our CI builds!

#  assert(StubRoutines::aarch64::complete() && Thread::current()->is_Compiler_thread() && Compile::current()->output() != __null) failed: not in output phase of C2 compilation

This change will need to be backed out.

**Comment by eastig** [comment]
> This is crashing our CI builds!
> 
> `# assert(StubRoutines::aarch64::complete() && Thread::current()->is_Compiler_thread() && Compile::current()->output() != __null) failed: not in output phase of C2 compilation`
> 
> This change will need to be backed out.

Hi David,
Do you have a stack trace?

**Comment by eastig** [comment]
@dholmes-ora, I created https://bugs.openjdk.org/browse/JDK-8291654

**Comment by eastig** [comment]
@dholmes-ora Fix removing the assert: https://github.com/openjdk/jdk/pull/9709

**Pull request closed without merging**